lib/engine-jit/src/engine.rs

```diff
@@ -193,130 +192,117 @@ impl JITEngineInner {
+            .chain(dynamic_function_trampolines.values())
+            .collect::<Vec<_>>();
+        let (executable_sections, data_sections): (Vec<_>, _) = custom_sections
+            .values()
+            .partition(|section| section.protection == CustomSectionProtection::ReadExecute);
+        self.code_memory.push(CodeMemory::new());
 
-        let mut allocated_function_call_trampolines: PrimaryMap<SignatureIndex, FunctionBodyPtr> =
-            PrimaryMap::new();
-        // let (indices, compiled_functions): (Vec<VMSharedSignatureIndex>, PrimaryMap<FunctionIndex, FunctionBody>) = function_call_trampolines.iter().map(|(sig_index, compiled_function)| {
-        //     let func_type = module.signatures.get(sig_index).unwrap();
-        //     let index = self.signatures.register(&func_type);
-        //     (index, compiled_function)
-        // }).filter(|(index, _)| {
-        //     !self.function_call_trampolines.contains_key(index)
-        // }).unzip();
-        for (sig_index, compiled_function) in function_call_trampolines.iter() {
-            let func_type = module.signatures.get(sig_index).unwrap();
-            let index = self.signatures.register(&func_type);
-            // if self.function_call_trampolines.contains_key(&index) {
-            //     // We don't need to allocate the trampoline in case
-            //     // it's signature is already allocated.
-            //     continue;
-            // }
-            let ptr = self
-                .code_memory
-                .allocate_for_function(registry, &compiled_function)
+        let (mut allocated_functions, allocated_executable_sections, allocated_data_sections) =
+            self.code_memory
+                .last_mut()
+                .unwrap()
+                .allocate(
+                    function_bodies.as_slice(),
+                    executable_sections.as_slice(),
+                    data_sections.as_slice(),
+                )
                 .map_err(|message| {
                     CompileError::Resource(format!(
-                        "failed to allocate memory for function call trampolines: {}",
+                        "failed to allocate memory for functions: {}",
                         message
                     ))
                 })?;
-            allocated_function_call_trampolines.push(FunctionBodyPtr(ptr));
-            let trampoline =
-                unsafe { std::mem::transmute::<*const VMFunctionBody, VMTrampoline>(ptr.as_ptr()) };
+
+        let allocated_functions_result = allocated_functions
+            .drain(0..functions.len())
+            .map(|slice| FunctionBodyPtr(slice as *mut [_]))
+            .collect::<PrimaryMap<LocalFunctionIndex, _>>();
+
+        let mut allocated_function_call_trampolines: PrimaryMap<SignatureIndex, FunctionBodyPtr> =
+            PrimaryMap::new();
+        for (sig_index, _) in function_call_trampolines.iter() {
+            let func_type = module.signatures.get(sig_index).unwrap();
```

Comment:
probably the same as `module.signatures[sig_index]` which is generally equivalent to `get(idx).unwrap()`